x-pack/plugins/spaces/server/lib/telemetry_client/telemetry_client.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ISavedObjectsRepository } from 'src/core/server';
+import { SPACES_TELEMETRY_TYPE } from '../../constants';
+import { CopyOptions, ResolveConflictsOptions } from '../copy_to_spaces/types';
+import { SpacesTelemetry, BooleanCount } from '../../model/spaces_telemetry';
+
+type IncrementCopySavedObjectsOptions = Pick<CopyOptions, 'createNewCopies' | 'overwrite'>;
+type IncrementResolveCopySavedObjectsErrorsOptions = Pick<
+  ResolveConflictsOptions,
+  'createNewCopies'
+>;
+
+const COPY_DEFAULT = Object.freeze({
+  total: 0,
+  createNewCopies: Object.freeze({ enabled: 0, disabled: 0 }),
+  overwrite: Object.freeze({ enabled: 0, disabled: 0 }),
+});
+const RESOLVE_COPY_ERRORS_DEFAULT = Object.freeze({
+  total: 0,
+  createNewCopies: Object.freeze({ enabled: 0, disabled: 0 }),
+});
+
+export class TelemetryClient {
+  constructor(
+    private readonly debugLogger: (message: string) => void,
+    private readonly repository: ISavedObjectsRepository
+  ) {}
+
+  public async getTelemetryData() {
+    this.debugLogger('getTelemetryData() called');
+    let spacesTelemetry: SpacesTelemetry = {};
+    try {
+      const result = await this.repository.get<SpacesTelemetry>(
+        SPACES_TELEMETRY_TYPE,
+        SPACES_TELEMETRY_TYPE
+      );
+      spacesTelemetry = result.attributes;
+    } catch (err) {
+      // do nothing
+    }
+    return spacesTelemetry;
+  }
+
+  public async incrementCopySavedObjects({
+    createNewCopies,
+    overwrite,
+  }: IncrementCopySavedObjectsOptions) {
+    const spacesTelemetry = await this.getTelemetryData();
+    const { apiCalls = {} } = spacesTelemetry;
+    const { copySavedObjects: current = COPY_DEFAULT } = apiCalls;
+
+    const attributes = {
+      ...spacesTelemetry,
+      apiCalls: {
+        ...apiCalls,
+        copySavedObjects: {
+          total: current.total + 1,
+          createNewCopies: incrementBooleanCount(current.createNewCopies, createNewCopies),
+          overwrite: incrementBooleanCount(current.overwrite, overwrite),
+        },
+      },
+    };
+    await this.updateTelemetryData(attributes);
+  }
```

Comment:
I feel like this class is doing the exact same thing as the `CoreTelemetryClient` except that it's tailored to the Spaces use case. 

Maybe instead of having many of these clients, each one implementing basically the same logic, we could just have one telemetry service in core that every plugin can use to add telemetry data:

As far as I can tell the logic is the following when we want to record usage data:
1. Retrieve saved object
3. Increment count (a bit like a redux reducer)
3. Update saved object

Something like this could encapsulate this logic:

```ts
// Plugin setup
const update = core.telemetryService.register(name, mappings);

// Route handler
update((current = defaults) => next);
```

e.g. for Spaces plugin:

```ts
// Plugin setup
const updateSpacesTelemetry = core.telemetryService.register('spaces-telemetry', mappings);

// Route handler
updateSpacesTelemetry((current = defaults) => ({
  total: current.total + 1,
  createNewCopies: incrementBooleanCount(current.createNewCopies, createNewCopies),
  overwrite: incrementBooleanCount(current.overwrite, overwrite),
}))
```

Just thinking out loud here, but this would make it a lot easier to add additional usage data. The only thing you'd need to implement and test as a plugin developer would be the reducer function.

Reply:
I like the idea and it sounds elegant, and I agree this is essentially duplicated code. However, I know we are trying hard to keep the core API surface as small as possible, and I'm not sure we want to add a whole service for something that we don't yet have additional use cases for. I'll defer to @elastic/kibana-core since they would be the owners of said service.